playground/backend/internal/cache/redis_cache.go

```diff
@@ -0,0 +1,148 @@
+	// ctx should be as a parameter
+	ctx := context.Background()
+
+	rc := RedisCache{redisClient: redis.NewClient(&redis.Options{Addr: addr})}
+	_, err := rc.redisClient.Ping(ctx).Result()
+	if err != nil {
+		return nil, err
+	}
+	return &rc, nil
+}
+
+func (rc *RedisCache) GetValue(pipelineId uuid.UUID, subKey SubKey) (interface{}, error) {
+	// ctx should be as a parameter
+	ctx := context.Background()
+
+	switch subKey {
+	case SubKey_Status:
+		var result pb.Status
+		err := get(ctx, rc.redisClient, pipelineId.String(), subKey, &result)
+		if err != nil {
+			return nil, err
+		}
+		return result, nil
+	default:
+		var result string
+		err := get(ctx, rc.redisClient, pipelineId.String(), subKey, &result)
+		if err != nil {
+			return nil, err
+		}
+		return result, nil
+	}
+}
+
+func (rc *RedisCache) SetValue(pipelineId uuid.UUID, subKey SubKey, value interface{}) error {
+	// ctx should be as a parameter
+	ctx := context.Background()
+
+	return set(ctx, rc.redisClient, pipelineId.String(), subKey, value)
+}
+
+func (rc *RedisCache) SetExpTime(pipelineId uuid.UUID, expTime time.Duration) error {
+	// ctx should be as a parameter
+	ctx := context.Background()
+
+	exists, err := rc.redisClient.Exists(ctx, pipelineId.String()).Result()
+	if err != nil {
+		return err
+	}
+	if exists == 0 {
+		return fmt.Errorf("key: %s doesn't exist", pipelineId)
+	}
+
+	rc.redisClient.Expire(ctx, pipelineId.String(), expTime)
+	return nil
+}
+
+// set puts value to Redis by key and subKey (key:subKey:value).
+// If it is the first value for this key then set expiration for key.
+func set(ctx context.Context, redisClient *redis.Client, key string, subKey SubKey, value interface{}) error {
+	subKeyMarsh, err := marshal(subKey)
+	if err != nil {
+		return err
+	}
+	valueMarsh, err := marshal(value)
+	if err != nil {
+		return err
+	}
+	_, err = redisClient.HSet(ctx, key, subKeyMarsh, valueMarsh).Result()
+	if err != nil {
+		return err
+	}
+
+	redisClient.Expire(ctx, key, defaultExpirationTime)
+	return nil
+}
+
+// get sets value by key and sybKey to result variable with converting data by subKey.
+// In case of variable doesn't exist returns error.
+func get(ctx context.Context, redisClient *redis.Client, key string, subKey SubKey, result interface{}) error {
+	marshSubKey, err := marshal(subKey)
+	if err != nil {
+		return err
+	}
+	value, err := redisClient.HGet(ctx, key, string(marshSubKey)).Result()
+	if err != nil {
+		return err
+	}
+	err = unmarshal([]byte(value), &result)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// marshal traverses the value v recursively.
+func marshal(data interface{}) ([]byte, error) {
+	result, err := json.Marshal(data)
+	if err != nil {
+		return nil, err
+	}
+	return result, nil
+}
+
+// unmarshal parses the data and stores the result
+func unmarshal(data []byte, result interface{}) error {
+	err := json.Unmarshal(data, result)
+	if err != nil {
+		return err
+	}
+	return nil
+}
```

Comment:
Do we really need to cover `json.Marshal` and `json.Unmarshal` with functions that just check errors since we also check these errors where we call `marshal` and `unmarshal`?

Reply:
Changed.